app/server/lib/requestUtils.ts

```diff
@@ -86,6 +88,9 @@ export function trustOrigin(req: Request, resp: Response): boolean {
   const origin = req.get('origin');
   if (!origin) { return true; } // Not a CORS request.
   if (process.env.GRIST_HOST && req.hostname === process.env.GRIST_HOST) { return true; }
+
+  if (isOwnInternalUrlHost(req.get('Host'))) { return true; }
+
```

Comment:
Besides the conflict with the changes I made in #859 (removing the `GRIST_HOST` exception above, and changing the type of `req`), I'm not sure what the intent of this exception is?
Something to keep in mind: every exception here is a potential vulnerability. Allowing a request based on the `Host` header, without even checking `Origin` seems strange.

Reply:
The intent is to allow requests made internally, assuming that the url of `APP_DOC_INTERNAL_URL` and `APP_HOME_INTERNAL_URL` cannot be resolved outside the internal network.

That being said, and as I understand that the `origin` header is passed along the `fetch` call made internally, I should check why it cannot pass the test of the below code.

Reply:
Do you have a test that fails without this exception?

Reply:
A manual test at least (duplicating the document), but with what I understand of the code, I have doubts, so I should try again without and see what happens.

Reply:
OK, I confirm I have this error if I remove this line:

> 18:12:42 Credentials not supported for cross-origin requests 
> Duplicate Document

This is because below, the `allowHost(req, new URL(origin))` checks whether the `Origin` header contains the same domain as for the `Host` header, because internally the `Origin` header is passed along from the request the client made (the one to copy) to the internal one (the one to fetch the document).

BTW, the justification (that I think is given below) puzzles me a lot :thinking::
https://github.com/gristlabs/grist-core/blob/3405ce902e6fbdcf5b95835a32091308ad173775/app/server/lib/Authorizer.ts#L686-L687

Reply:
I proposed this alternative solution, though I have doubts about the implications, even if I *tend* to think that would be OK:
https://github.com/gristlabs/grist-core/pull/915/commits/93be827098956a47decd78139eb5c0bd0033d156

Have you got any thoughts?

Reply:
Some background on this comment:
```
// Pass along the original Origin since it may 
// play a role in granular access control.
```
The origin is made available in access rules: https://support.getgrist.com/access-rules/#access-rule-conditions
This is an unfinished and inconsistently available feature though. The thought was to use it for a common weak piece of security some services offer, where you can set something to only be usable from a particular site and not from other sites (in the context of browser usage). Nobody was asking for that, and I don't know of anyone doing that, so if this wrinkle is in the way of anything, I think we could cut it out.

Reply:
@fflorent what you found matches my own research. I had likewise concluded that preserving the `Host` header in some form when forwarding requests would be required, as access decisions in the Grist codebase seem to often be made by comparing `Origin` to `Host` (as an aside, here's an unfortunate detail : the [`getOrigin`]( https://github.com/gristlabs/grist-core/blob/dd83b7f67830f032fcd2cae8de1341c4d88cf32f/app/server/lib/requestUtils.ts#L328-L339) function in `requestUtils` actually returns a value based on the `Host` header, not the `Origin` header).

Also, these decisions can be quite involved, and may even require a database access as in the case of custom domains — which is unfortunately a code path that as far as I know no one is exercising apart from Grist SaaS, which makes me wary of suggesting any changes to that logic.

Your proposal that replicates the `Host` header in `getTransitiveHeaders` does go in the same direction I had in mind. But I'm afraid it may break some deployments, as the `Host` header can play a critical role in dispatching an HTTP request. Here's a contrived example that shows how it could break:
- Given a Grist instance with a public URL of `https://grist.example.org`;
- Handling an incoming user request at `https://grist.example.org/…` (therefore, `Host` is `grist.example.org`) that triggers a forwarded request to e.g. `http://grist-worker.internal`;
- Copying the `Host` header from the public request to the forwarded request would result in making a request to `http://grist-worker.internal` with a `Host` header value of `grist.example.org` (as opposed to a `Host` value of `grist-worker.internal` as is currently automatically generated by the HTTP client, i.e. `node-fetch` as far as I can tell).
Depending on how the HTTP server on `grist-worker.internal` is setup, this could cause the request to fail. If `grist-worker.internal` resolves directly to a Grist worker, it's likely to work as the Node.JS HTTP server basically ignores the incoming `Host` value. But if for some reason a reverse proxy/load balancer is in front of that server, it may reject the request or route it to the wrong server.

This very common problem with reverse proxying is generally addressed by copying the `Host` header value to another header such as `X-Forwarded-Host` instead, so that the code handling the request can make decisions based on the `Host` value sent by the initial client, without compromising request routing. Of course, this implies careful handling of whether to trust such a header, since it can easily be spoofed by a client if it is not filtered at the reverse proxy (see e.g. @dsagal 's remark [here](https://github.com/gristlabs/grist-core/pull/859#discussion_r1537005097) about the related `X-Forwarded-Proto` header).

Reply:
@paulfitz the use of `Origin` in access rules is another wrinkle I hadn't discovered yet, thanks for bringing it to our attention.

If I understand what you're suggesting, removing the `Origin` header from forwarded requests would likely let them more easily through, as requests without an `Origin` header are considered same-origin and exempted from any further checks in [`trustOrigin`](https://github.com/gristlabs/grist-core/blob/dd83b7f67830f032fcd2cae8de1341c4d88cf32f/app/server/lib/requestUtils.ts#L89).

It seems to me though there are other parts of the code that currently match the `Host` header with e.g. parts of the session cookie, such as [this code in `Authorizer`](https://github.com/gristlabs/grist-core/blob/dd83b7f67830f032fcd2cae8de1341c4d88cf32f/app/server/lib/Authorizer.ts) — again, this appears to be custom host-specific so difficult to exercise outside of Grist SaaS. Correctly forwarding `Host` would probably help such code work, even though it doesn't seem to pose a problem currently because it probably isn't hit by any forwarded requests, but that seems rather an accident.

In general, it seems to me that if any header (such as `Origin` or `Cookie`) is forwarded, any value that it may be checked against (here, `Host`) needs to be made available as well to preserve the full security context.

Reply:
@fflorent In any case, I strongly suggest you rebase your branch onto `main` as the code in question was affected by the merging of #859 .

Reply:
> This very common problem with reverse proxying is generally addressed by copying the Host header value to another header such as X-Forwarded-Host instead, so that the code handling the request can make decisions based on the Host value sent by the initial client, without compromising request routing. Of course, this implies careful handling of whether to trust such a header, since it can easily be spoofed by a client if it is not filtered at the reverse proxy (see e.g. @dsagal 's remark https://github.com/gristlabs/grist-core/pull/859#discussion_r1537005097 about the related X-Forwarded-Proto header).

Could we take advantage of the [express `'trust proxy'`](https://expressjs.com/en/guide/behind-proxies.html) feature? We could introduce another env variable to set IP addresses / subnets written in CSV format, and document it as dangerous unless the `X-Forwarded-For`, `X-Forwarded-Host`, and `X-Forwarded-Proto` are not cleared/overwritten by the RP.